tpm2/crypt.go

```diff
@@ -17,17 +17,90 @@
  *
  */
 
-package secboot
+package tpm2
 
 import (
+	"bufio"
+	"bytes"
 	"errors"
 	"fmt"
 	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
 
 	"golang.org/x/xerrors"
+
+	"github.com/snapcore/secboot"
+	"github.com/snapcore/secboot/internal/luks2"
+)
+
+var (
+	luks2Activate                        = luks2.Activate
+	secbootActivateVolumeWithRecoveryKey = secboot.ActivateVolumeWithRecoveryKey
 )
 
-func unsealKeyFromTPM(tpm *TPMConnection, k *SealedKeyObject, pin string) ([]byte, error) {
+// XXX: This code is duplicated temporarily from github.com/snapcore/secboot:crypt.go
+// It will go away once https://github.com/snapcore/secboot/pull/128/ lands and this
+// code has been ported to use it.
+type execError struct {
+	path string
+	err  error
+}
+
+func (e *execError) Error() string {
+	return fmt.Sprintf("%s failed: %s", e.path, e.err)
+}
+
+func (e *execError) Unwrap() error {
+	return e.err
+}
+
+func wrapExecError(cmd *exec.Cmd, err error) error {
+	if err == nil {
+		return nil
+	}
+	return &execError{path: cmd.Path, err: err}
+}
+
+// XXX: This code is duplicated temporarily from github.com/snapcore/secboot:crypt.go
+// It will go away once there is an abstract interface for handling authorization requests,
+// or we figure out a way to do activation with TPM key files using the new API so that
+// this one can be removed.
+func askPassword(sourceDevicePath, msg string) (string, error) {
+	cmd := exec.Command(
+		"systemd-ask-password",
+		"--icon", "drive-harddisk",
+		"--id", filepath.Base(os.Args[0])+":"+sourceDevicePath,
+		msg)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return "", wrapExecError(cmd, err)
```

Comment:
just noticed in passing that this code path is not tested, also not for the original copy of the code, but I also wonder if we still need execError as isExecError is not used anymore?

Reply:
It looks like execError is not needed now.